### PR TITLE
feat: Make artifacts discoverable in the DAG. Fixes #8494

### LINKED ISSUE #8494: Artifact UI: Make artifacts discoverable in UI
It is currently very hard to find artifacts in the UI. Make that easier.

### PR DESCRIPTION
**Pull request by alexec**
Signed-off-by: Alex Collins <alex_collins@intuit.com>

Fixes #8494

This PR allows users, if they want, to show artifacts in the DAG. They can toggle this on and off:

image

Their preference is set in local storage so it remains even if the browser is closed.

Input and output artifacts are shown in the DAG:

image

When the user clicks on them, name and a download link is shown:

image

This is not much by itself, but once we have the option to serve artifacts, it will allow us the panel to show the artifact in-browser.

This is what is looks like with artifacts disable.

image

### CODE REVIEW

**Comment by alexec** [comment]
image

image

image

image

**Comment by alexec** [comment]
Adding lots of reviewers to try and get different viewpoints.

File: ui/src/app/shared/components/first-time-user-panel.tsx [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import * as React from 'react';
+import {useEffect, useState} from 'react';
+
+// This wraps the children components in a FTU which allows you to provide an explanation of how to use the feature.
+// The child components are hidden behind the panel until the users dismissed the panel by clicking "Continue".
+// The fact the users dismissed the panel is persisted in local storage and the panel not show again.
+export const FirstTimeUserPanel = ({id, explanation, children}: {id: string; explanation: string; children: React.ReactElement}) => {
```
Comment (terrytangyuan) [inline ui/src/app/shared/components/first-time-user-panel.tsx:7]: Perhaps move this to a separate PR?

File: ui/src/app/shared/components/give-feedback-link.tsx [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import * as React from 'react';
+
+export const GiveFeedbackLink = ({href}: {href: string}) => (
```
Comment (terrytangyuan) [inline ui/src/app/shared/components/give-feedback-link.tsx:3]: I feel this and the first user panel should be a separate PR

File: ui/src/app/workflows/components/workflow-details/artifact-panel.tsx [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    archived
+}: {
+    workflow: Workflow;
+    artifact: Artifact & {nodeId: string; artifactNameDiscriminator: string};
+    archived?: boolean;
+}) => {
+    const downloadUrl = services.workflows.getArtifactDownloadUrl(workflow, artifact.nodeId, artifact.name, archived, artifact.artifactNameDiscriminator === 'input');
+
+    const key = artifactKey(artifact)
+        .split('/')
```
Comment (terrytangyuan) [inline ui/src/app/workflows/components/workflow-details/artifact-panel.tsx:26]: Is there a system agnostic way to do this?

File: ui/src/app/shared/components/first-time-user-panel.tsx (hunk 1 above)
Comment (alexec) [reply ui/src/app/shared/components/first-time-user-panel.tsx:7]: Kinda of a small PR. Do you mean before or afterwards?

Comment (terrytangyuan) [reply ui/src/app/shared/components/first-time-user-panel.tsx:7]: It's up to you. It's just for narrowing down the changes in this PR to make it easier for others to review but I don't have a strong opinion on moving them to a different PR.

File: ui/src/app/workflows/components/workflow-details/artifact-panel.tsx (hunk 3 above)
Comment (alexec) [reply ui/src/app/workflows/components/workflow-details/artifact-panel.tsx:26]: these are S3 keys, so always use `/` not `\\`?

**Review by terrytangyuan — APPROVED** [review]
Great work! LGTM from me. This would be super useful for our users.
